### PR TITLE
remove unnecessary Avx512VL ISA flags.

### PR DESCRIPTION
**Pull request by Ruihan-Yin**
More context in https://github.com/dotnet/runtime/pull/103019#discussion_r1625346678

To save some space in `XArchIntrinsicConstants`, this PR turns `Avx512*_VL` ISA flags per  subset into a converged flag: `Avx512F_VL`, such that we can hold more ISAs within this enum, this will be the pre-work for the ongoing APX project.

### CODE REVIEW

File: src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs [hunk 1]
```diff
@@ -73,17 +73,13 @@ private static class XArchIntrinsicConstants
             public const int Avx512f = 0x8000;
             public const int Avx512f_vl = 0x10000;
```
Comment (tannergooding) [inline src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs:74]: Can we just have this be `Avx512vl`, as that's the actual CPUID feature name and matches the casing we use elsewhere?

File: src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs [hunk 2]
```diff
@@ -73,17 +73,13 @@ private static class XArchIntrinsicConstants
...
-            public const int Avx10v1 = 0x4000000;
-            public const int Avx10v1_v256 = 0x8000000;
-            public const int Avx10v1_v512 = 0x10000000;
+            public const int Avx512cd = 0x40000;
+            public const int Avx512dq = 0x80000;
+            public const int Avx512Vbmi = 0x100000;
+            public const int Serialize = 0x200000;
+            public const int Avx10v1 = 0x400000;
+            public const int Avx10v1_v256 = 0x800000;
+            public const int Avx10v1_v512 = 0x1000000;
```
Comment (tannergooding) [inline src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs:82]: It'd be nice to remove these as well.

The former is unnecessary and the latter is implied by `Avx512F` existing.

File: src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs (hunk 1 above)
Comment (Ruihan-Yin) [reply src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs:74]: I agree, will rename this part accordingly.

File: src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs (hunk 2 above)
Comment (Ruihan-Yin) [reply src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs:82]: I can make this change, and just make sure, does this part have conflicting changes with the ongoing Avx10 PR? I know some changes happening there for this as well, but didn't follow closely.

Comment (tannergooding) [reply src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs:82]: That PR is removing `Avx10v1_v256`, so removing it here as well should help avoid a conflict.

**Comment by jkotas** [comment]
You will also need to fix [this](https://github.com/dotnet/runtime/blob/f6a7ebbb81540401e6b5520afa3ba87c2bd6bcfe/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/HardwareIntrinsicHelpers.Aot.cs#L55-L56) to check that all feature bits are present for features described by multiple bits.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/HardwareIntrinsicHelpers.Aot.cs [hunk 3]
```diff
@@ -49,12 +49,46 @@ public static MethodIL EmitIsSupportedIL(MethodDesc method, FieldDesc isSupporte
...
+                case InstructionSet.X64_AVX512F_VL:
+                case InstructionSet.X64_AVX512F_VL_X64:
+                case InstructionSet.X64_AVX512BW_VL:
+                case InstructionSet.X64_AVX512BW_VL_X64:
+                case InstructionSet.X64_AVX512CD_VL:
+                case InstructionSet.X64_AVX512CD_VL_X64:
+                case InstructionSet.X64_AVX512DQ_VL:
+                case InstructionSet.X64_AVX512DQ_VL_X64:
+                case InstructionSet.X64_AVX512VBMI_VL:
+                case InstructionSet.X64_AVX512VBMI_VL_X64:
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/HardwareIntrinsicHelpers.Aot.cs:65]:
```suggestion
            if (!uint.IsPow2((uint)flag))
            {
```
You should be able to just check whether the flag has one or more bits.

File: src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs (hunk 2 above)
Comment (Ruihan-Yin) [reply src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs]: I saw Avx10v1_V256 was removed from InstructionSetDesc.txt, so what we want here is don't even define `Avx10v1_V256` and `Avx10v1_V512`, and rely on `EVEX` to make sure JIT backend will correctly handle Avx10 nodes? Just to confirm.

Comment (tannergooding) [reply src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs]: Right. `Avx10v1_v256` is just part of `Avx10v1`, so it's unnecessary.

Likewise, `Avx10v1_v512` is really just `Avx10v1 + Avx512`, so doing those checks instead fits the need and saves us bits.

Comment (Ruihan-Yin) [reply src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs]: Thanks for the inputs,

There are duplicated implications there: https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/Common/JitInterface/ThunkGenerator/InstructionSetDesc.txt#L141

Are they intended to be like this, or this shall be removed, I can make the changes if needed.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/HardwareIntrinsicHelpers.Aot.cs [hunk 4]
```diff
@@ -49,12 +49,28 @@ public static MethodIL EmitIsSupportedIL(MethodDesc method, FieldDesc isSupporte
...
+            if(!uint.IsPow2((uint)flag))
+            {
+                // These are the ISAs managed by multiple-bit flags.
+                // we need to emit different IL to handle the checks.
+                // (isSupportedField & flag) == flag
+                codeStream.Emit(ILOpcode.ldsfld, emit.NewToken(isSupportedField));
+                codeStream.EmitLdc(flag);
+                codeStream.Emit(ILOpcode.and);
+                codeStream.EmitLdc(flag);
+                codeStream.Emit(ILOpcode.beq);
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/HardwareIntrinsicHelpers.Aot.cs:61]:
```suggestion
                codeStream.Emit(ILOpcode.ceq);
```
`beq` is "branch on equal". This is needs to be compare instead.

Comment (Ruihan-Yin) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/HardwareIntrinsicHelpers.Aot.cs]: Corrected, thanks for pointing out!

**Comment by Ruihan-Yin** [comment]
rebased the branch to resolve conflicts. 

I kept `Avx10v1_V512` as an ISA to make it consistent with the node definition in `hwintrinsicxarchlist.h`. If we need further changes, I can try to make it within this PR.

The fails look like connection issue.

File: src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs (hunk 2 above)
Comment (tannergooding) [reply src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs]: > Are they intended to be like this, or this shall be removed, I can make the changes if needed.

No and I'm actually fixing that in https://github.com/dotnet/runtime/pull/103241

> It looks fine to just remove it from XArchIntrinsicConstants but keep the definition as an ISA, and as mentioned, set InstructionSet_Avx10v1_V512 based on XArchIntrinsicConstants_Avx10v1 and XArchIntrinsicConstants_Avx512f

Right, this is what I had meant should happen.

Any refactoring that changes how `InstructionSet_*` works might be possible but would be a more involved change.

Comment (Ruihan-Yin) [reply src/coreclr/tools/Common/Compiler/HardwareIntrinsicHelpers.cs]: Thanks for the explanation, in that case, I can wait for this PR go in first, and I will handle the conflict accordingly.

**Comment by Ruihan-Yin** [comment]
Hi @tannergooding, 

Just making sure, I am looking at the #103241, in that PR, we have folded all the Avx512 bits into 1, is that expected moving on, or we are still looking for separate them out to be `F+VL+BW+CD+DQ` as discussed [here](https://github.com/dotnet/runtime/pull/103019#discussion_r1625346678). If not, I suppose the only thing left for this PR would be removing Avx10v1_V512?

File: src/native/minipal/cpufeatures.c [hunk 5]
```diff
@@ -263,13 +263,6 @@ int minipal_getcpufeatures(void)
                                         {
                                             result |= XArchIntrinsicConstants_Evex;
                                             result |= XArchIntrinsicConstants_Avx10v1;
-
-                                            bool isV512Supported = (cpuidInfo[CPUID_EBX] & (1 << 18)) != 0;     // Avx10/V512
```
Comment (jkotas) [inline src/native/minipal/cpufeatures.c:267]: Should we assert that the `Avx10/V512` bit is true iff `XArchIntrinsicConstants_Avx512` is set?

Comment (Ruihan-Yin) [reply src/native/minipal/cpufeatures.c:267]: We are relying on this [part](https://github.com/dotnet/runtime/pull/103144/files#diff-eba17a72b6caa8b52b196f3b84bfd20dfeaf39a89f410943a692d1f33f569f3bR1437) to set `Avx10/V512`, now `Avx10/V512` will be set iff `EVEX`, `Avx10v1` and `Avx512` all present.

**Review by tannergooding — APPROVED** [review]

File: src/native/minipal/cpufeatures.c [hunk 6]
```diff
@@ -263,13 +263,6 @@ int minipal_getcpufeatures(void)
                                         {
                                             result |= XArchIntrinsicConstants_Evex;
                                             result |= XArchIntrinsicConstants_Avx10v1;
```
Comment (jkotas) [inline src/native/minipal/cpufeatures.c:265]:
```suggestion
                                            result |= XArchIntrinsicConstants_Avx10v1;
                                            
                                            // We assume that the Avx10/V512 support can be inferred from
                                            // both Avx10v1 and Avx512 being present.
                                            assert(((cpuidInfo[CPUID_EBX] & (1 << 18)) != 0) ==            // Avx10/V512
                                                ((result & XArchIntrinsicConstants_Avx512) != 0));
```

File: src/native/minipal/cpufeatures.c (hunk 5 above)
Comment (jkotas) [reply src/native/minipal/cpufeatures.c:267]: Right, I meant to document this via an assert. Left a suggestion.

Comment (Ruihan-Yin) [reply src/native/minipal/cpufeatures.c:267]: Thanks for the suggestion!

**Comment by Ruihan-Yin** [comment]
Except known failures, the build fail looks like due to timeout, and not related the changes.

**Comment by tannergooding** [comment]
Rerunning CI to get it in a clean state

**Comment by Ruihan-Yin** [comment]
Build Analysis has shown green, PR should be in clean state.

**Comment by tannergooding** [comment]
@jkotas, did you have any other feedback here or can I merge once CI is showing green?

**Comment by jkotas** [comment]
LGTM

**Comment by Ruihan-Yin** [comment]
image
I was not able to reproduce the CI fail locally.

From the call stack, it is more likely to be connection issue.

**Comment by Ruihan-Yin** [comment]
Thanks for all the suggestions and help!

**Pull request merged**